User Service 테스트 생성 및 비즈니스 로직 생성

seoko97 opened the pull request:
* Closes #21
* Closes #22

## ✨ **구현 기능 명세**
- User Service 관련 테스트 코드 및 비즈니스 로직을 생성합니다.
- 구현된 함수: `getUser`, `createUser`
- `getUser`: parameter `id` 값에 따른 user정보를 호출합니다.
- `createUser`: parameter `userInfo { username, email, profileImage }`를 통해 user를 생성합니다.

## 🎁 **주목할 점**
- 각 함수에 대한 error 처리 세분화

## 😭 **어려웠던 점**
- userService에 관한 테스트 로직 중 기존 구현된 함수에 대한 Stup을 제공하는 중 메소드 이름을 항상 작성해야되서 에러를 해결하는데 시간이 걸렸습니다.


## 🌄 **스크린샷**
- `user.service.spec.ts`: 총 2개의 test가 정상적으로 동작
![image](https://user-images.githubusercontent.com/60173534/226282669-787b45a8-9dd7-4f4f-8068-75a0ca1424a4.png)

seoko97 committed:
test.user mock data 생성 (#21)(#18)

seoko97 committed:
test: user service 테스트 코드 작성 (#21)
- getUser, createUser 테스트 코드 작성

seoko97 committed:
feat: createUserDTO 변경 및 적용 (#22)
- 기존 username에서 email, profileImage 추가

seoko97 committed:
feat: 사용하지 않는 모듈 제거

seoko97 committed:
feat: user service 로직 추가 (#22)
- getUser: 유저 정보 조회
- createUser: 유저 생성
- 각 로직 별 에러 처리

seoko97 committed:
feat: user repository custom method 추가 (#21, #22)
- getUserById : id로 유저 정보 조회
- 반복적으로 사용할 것 같은 메소드들을 추가하였습니다.

pyo-sh commented on test/api/user/mock.ts:
```diff
@@ -0,0 +1,12 @@
+import { User } from '@/entity/user.entity';
+
+const USER_MOCK: User = {
+  id: 1,
+  email: 'test@test.com',
+  username: '테스트',
+  profileImage: '이미지',
+  createdAt: new Date('2023-03-07 15:38:06.785155'),
+  updatedAt: new Date('2023-03-07 15:38:06.785155'),
+};
```
제가 이전에 테스트 관련해서 예시코드 작성할 때에는 Test Double에 대한 지식이 없었나봅니다 😓
이런식으로 User 값을 테스트 시 실체화해서 뱉는 것 자체가 Stub 이라고 하네요!
네이밍이 `USER_STUB` 과 같은 방식으로 바뀌는건 어떤가요?

seoko97 replied:
확인했습니다! 관련 코드도 stub으로 바꿀게요

pyo-sh commented on src/api/user/user.service.ts:
```diff
@@ -1,43 +1,40 @@
+        where: { email: userInfo.email },
+      });
+
+      if (user) {
+        new BadRequestException('이미 존재하는 유저입니다.');
       }
+
+      return this.userRepo.createUser(userInfo);
+    } catch (error) {
+      new InternalServerErrorException('서버 에러입니다.');
     }
```
유닛 테스팅을 하면 항상 힘든점인 것 같습니다.
Service 로직에서 response로 넘겨줄 Exception들을 선언했을 때 **Exception에 대한 테스팅은 Service에서 해야하나 Controller에서 해야하는가?** 에 대한 질문입니다.
    - Controller에서 supertest를 통해 res, req 값을 테스팅해야한다는 점
    - Service에서 Exception을 구현하므로 try~catch로 테스팅 할 수 있다는 점

관련해서 논의해보고 싶습니다!

jintak0401 replied:
개인적인 생각으로는 Service에서 실행해주어야 할 것 같습니다. [Controller와 Service의 차이](https://geonoo.tistory.com/152)를 생각해보았을 때, controller 역할은 request를 받고 service에게 요청하고 응답받기, service는 controller부터 호출되어 적절한 로직으로 응답을 도출하기입니다. 

"esponse의 예외"라는 것은 비즈니스 로직 상에 해당된다고 생각합니다. Contoller가 응답으로 불가능한 값을 판단하기 보다는, Service에서 해주는 것이 더 맞다고 생각해요

seoko97 replied:
Timegambit님 의견에 동의합니다. Controller를 비즈니스 로직의 집합으로 생각했을때, Service단에서 비즈니스 로직을 정의하며 발생하는 에러들을 생각해 예외상황을 정의하여 에러를 반환해야된다고 생각합니다. 반박 받습니다.

jintak0401 reviewed (approved):
고생하셨습니다

seoko97 committed:
fix: throw error 추가 (#22) (#23)
- error 시 throw문을 추가하지 않아 에러가 발생하지 않던 문제 해결

seoko97 committed:
test: user service reject시 test case 추가 (#21) (#23)
- getUser: 유저가 존재하지 않을 때 "존재하지 않는 유저입니다." 반환
- createUser: 유저가 이미 존재할 때 "이미 존재하는 유저입니다." 반환

seoko97 committed:
feat: mock -> stub 으로 변경 (#23)
- mock: 특정 메서드를 재대로 콜하는지
- stub: 더미 객체를 생성하고 실제로 동작하는 것처럼 보이게 하만든 각자 객체

seoko97 committed:
feat: 변경하지 않은 변수 추가 (#23)

pyo-sh reviewed (approved):
고생이 많으십니다...
피드백 수용에 Error 부분 처리까지 테스팅 너무 좋습니다~~
